grpc/src/main/java/com/linecorp/armeria/common/grpc/GsonGrpcJsonMarshallerBuilder.java

```diff
@@ -32,40 +33,98 @@
+                    (Function<JsonFormat.Parser, JsonFormat.Parser>) jsonParserCustomizer;
+            this.jsonParserCustomizer = cast;
+        } else {
+            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(jsonParserCustomizer);
+        }
+        return this;
+    }
+
     /**
      * Adds a {@link Consumer} that can customize the {@link JsonFormat.Parser}
      * used when deserializing a JSON payload into a {@link Message}.
+     *
+     * @deprecated {@link JsonFormat.Parser} is immutable so all changes applied in the {@link Consumer}
+     *     will be lost. Please use the {@link #jsonParserCustomizer(Function) jsonParserCustomizer}
+     *     which accepts {@link Function} parameter instead.
      */
+    @Deprecated
     public GsonGrpcJsonMarshallerBuilder jsonParserCustomizer(
             Consumer<? super JsonFormat.Parser> jsonParserCustomizer) {
         requireNonNull(jsonParserCustomizer, "jsonParserCustomizer");
         if (this.jsonParserCustomizer == null) {
             @SuppressWarnings("unchecked")
             final Consumer<JsonFormat.Parser> cast = (Consumer<JsonFormat.Parser>) jsonParserCustomizer;
-            this.jsonParserCustomizer = cast;
+            this.jsonParserCustomizer = parser -> {
+                cast.accept(parser);
+                return parser;
+            };
         } else {
-            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(jsonParserCustomizer);
+            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(parser -> {
+                jsonParserCustomizer.accept(parser);
+                return parser;
+            });
         }
         return this;
     }
 
     /**
      * Adds a {@link Consumer} that can customize the {@link JsonFormat.Printer}
      * used when serializing a {@link Message} into a JSON payload.
+     *
+     * @deprecated {@link JsonFormat.Printer} is immutable so all changes applied in the {@link Consumer}
+     *     will be lost. Please use the {@link #jsonPrinterCustomizer(Function) jsonParserCustomizer}
+     *     which accepts {@link Function} parameter instead.
      */
+    @Deprecated
     public GsonGrpcJsonMarshallerBuilder jsonPrinterCustomizer(
             Consumer<? super JsonFormat.Printer> jsonPrinterCustomizer) {
         requireNonNull(jsonPrinterCustomizer, "jsonPrinterCustomizer");
         if (this.jsonPrinterCustomizer == null) {
             @SuppressWarnings("unchecked")
             final Consumer<JsonFormat.Printer> cast = (Consumer<JsonFormat.Printer>) jsonPrinterCustomizer;
+            this.jsonPrinterCustomizer = printer -> {
+                cast.accept(printer);
+                return printer;
+            };
+        } else {
+            this.jsonPrinterCustomizer = this.jsonPrinterCustomizer.andThen(printer -> {
+                jsonPrinterCustomizer.accept(printer);
+                return printer;
+            });
+        }
+        return this;
```

Comment:
Ditto - we could do something similar to the above comment here for `jsonPrinterCustomizer(Consumer)`.

Reply:
added logging here as well https://github.com/line/armeria/pull/6146/commits/0812f5a2c828b8f5878b40d2c5a3b6e93a86ddee